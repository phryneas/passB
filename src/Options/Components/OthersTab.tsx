import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Typography,
} from 'material-ui';
import {Cancel, Check, Clear} from 'material-ui-icons';
import * as React from 'react';
import {Interfaces, Symbols} from 'Container';
import {lazyInject} from 'Decorators/lazyInject';

interface State {
  showResetDialog: boolean;
}

export class OthersTab extends React.Component<{}, State> {
  public state: State = {
    showResetDialog: false,
  };

  @lazyInject(Symbols.PassB)
  private passB: Interfaces.PassB;

  public render(): JSX.Element {
    const {
      showResetDialog,
    } = this.state;

    return (
      <div>
        <List>
          <ListItem
            button={true}
            onClick={() => this.setState({showResetDialog: true})}
          >
            <ListItemIcon><Clear/></ListItemIcon>
            <ListItemText primary={browser.i18n.getMessage('options_reset_extension')}/>
          </ListItem>
        </List>
        <Dialog
          onClose={() => this.setState({showResetDialog: false})}
          maxWidth="xs"
          open={showResetDialog}
        >
          <DialogContent>
            <Typography type="subheading">{browser.i18n.getMessage('options_reset_extension_confirm')}</Typography>
          </DialogContent>
          <DialogActions>
            <Button onClick={() => this.setState({showResetDialog: false})} color="primary">
              <Cancel/>
            </Button>
            <Button
              onClick={() => {
                browser.storage.local.clear();
                this.passB.reloadExtension();
                browser.tabs.getCurrent().then((tab: browser.tabs.Tab) => {
                  if (tab && tab.id) {
                    browser.tabs.remove(tab.id);
                  }
                });
              }}
              color="primary"
            >
              <Check/>
            </Button>
          </DialogActions>
        </Dialog>
      </div>
    );
  }
}
