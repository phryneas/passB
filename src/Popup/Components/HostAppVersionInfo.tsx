import {Button, List, ListItem, ListItemText, Popover} from 'material-ui';
import {Warning} from 'material-ui-icons';
import * as React from 'react';
import {connect} from 'react-redux';
import {lt, valid} from 'semver';
import {latestHostAppVersion} from 'Constants';
import {StoreContents} from 'InjectableInterfaces/State';
import {getVersion} from 'State/HostApp/index';

export interface Props {
  visible: boolean;
  hostAppVersion: string;
}

export interface State {
  open: boolean;
}

export class HostAppVersionInfoComponent extends React.Component<Props, State> {
  public state: State = {
    open: false,
  };

  private button: HTMLButtonElement;

  public render(): JSX.Element[] | false {
    const {visible, hostAppVersion} = this.props;
    const {open} = this.state;

    if (!visible) {
      return false;
    }

    return [(
      <Button
        rootRef={(button: HTMLButtonElement) => this.button = button}
        key="btn"
        color="accent"
        onClick={() => this.setState({open: true})}
      >
        <Warning/>
      </Button>
    ),      (
      <Popover
        key="popover"
        anchorEl={this.button}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        open={open}
        onClose={() => this.setState({open: false})}
      >
            <List>
              <ListItem>
                <ListItemText primary={browser.i18n.getMessage('popup_host_app_nag_update', [hostAppVersion, latestHostAppVersion])} />
              </ListItem>
              <ListItem
                button={true}
                className="hostAppInfo"
                onClick={() => browser.tabs.create({url: 'https://passb.github.io/host_application.html'})}
              >
                <ListItemText primary={browser.i18n.getMessage('popup_host_app_error_option_install')}/>
              </ListItem>
            </List>
      </Popover>
    )];
  }
}

export const mapStateToProps = (state: StoreContents): Props => {
  const version = getVersion(state);
  return {
    hostAppVersion: version,
    visible: valid(version) !== null && lt(version, latestHostAppVersion),
  };
};

export const HostAppVersionInfo = connect(mapStateToProps)(HostAppVersionInfoComponent);
