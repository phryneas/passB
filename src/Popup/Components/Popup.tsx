import {AppBar, Button, Card, CardContent, Toolbar, Typography, WithStyles} from 'material-ui';
import {Refresh, Settings} from 'material-ui-icons';
import {withStyles, StyleRules} from 'material-ui/styles';
import * as React from 'react';
import Tab = browser.tabs.Tab;
import {MemoryRouter, Route, RouteComponentProps, RouteProps, Switch} from 'react-router';
import {Interfaces, Symbols} from 'Container';
import {lazyInject} from 'Decorators/lazyInject';
import {ErrorBoundary} from './ErrorBoundary';
import {HostAppErrorWrapper} from './HostAppErrorWrapper';
import {HostAppVersionInfo} from './HostAppVersionInfo';
import {ListView} from './Views/ListView';

interface Props {
}

interface State {
  activeTab?: Tab;
}

const styles: StyleRules = {
  flexLeft: {
    flex: 1,
    marginLeft: '15px',
  },
  cardSize: {
    maxHeight: '390px',
    overflow: 'auto',
  },
};

class ClassLessPopup extends React.Component<Props & WithStyles<keyof typeof styles>, State> {
  public state: State = {};
  private gatheredRoutes: RouteProps[] = this.gatherRoutes();

  @lazyInject(Symbols.PassB)
  private passB: Interfaces.PassB;

  public componentDidMount(): void {
    browser.tabs.query({
      active: true,
      currentWindow: true,
    }).then((tabs: Tab[]) => this.setState({activeTab: tabs[0]}));
  }

  public render(): JSX.Element {
    const {activeTab} = this.state;
    const {classes} = this.props;

    return (
      <MemoryRouter>
        <div>
          <AppBar position="fixed">
            <Toolbar>
              <Typography type="title" color="inherit" className={classes.flexLeft}>
                {browser.i18n.getMessage('extensionName')}
              </Typography>
              <HostAppVersionInfo />
              <Button
                color="contrast"
                onClick={() => {
                  this.passB.reloadEntries();
                  window.close();
                }}
              >
                <Refresh/>
              </Button>
              <Button
                color="contrast"
                onClick={() => {
                  browser.runtime.openOptionsPage();
                  window.close();
                }}
              >
                <Settings/>
              </Button>
            </Toolbar>
          </AppBar>
          <Card>
            <CardContent className={classes.cardSize}>
              <ErrorBoundary>
                <HostAppErrorWrapper>
                  <Switch>
                    {this.gatheredRoutes.map((route: RouteProps) => <Route key={String(route.path)}  {...route} />)}
                    <Route
                      render={
                        ({history}: RouteComponentProps<{}>) => (
                          <ListView
                            url={activeTab && activeTab.url ? activeTab.url : ''}
                          />
                        )}
                    />
                  </Switch>
                </HostAppErrorWrapper>
              </ErrorBoundary>
            </CardContent>
          </Card>
        </div>
      </MemoryRouter>
    );
  }

  private gatherRoutes(): RouteProps[] {
    const gatheredRoutes: RouteProps[] = [];
    for (const extension of this.passB.getAllExtensions()) {
      for (const route of extension.routes) {
        if (!(route.path || '').startsWith(`/extension/${extension.name}/`)) {
          throw Error(
            `every route path for extension ${extension.name} has to start with "/extension/${extension.name}/"`,
          );
        }
        gatheredRoutes.push(route);
      }
    }

    return gatheredRoutes;
  }
}

export const Popup = withStyles<keyof typeof styles>(styles)(ClassLessPopup);
