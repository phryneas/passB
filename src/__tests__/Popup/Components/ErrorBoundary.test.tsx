import {shallow, ShallowWrapper} from 'enzyme';
import * as React from 'react';
import {RouteComponentProps} from 'react-router';
import {getMockRouteComponentProps} from '__test_helpers__/getMockRouteComponentProps';
import {ErrorBoundaryComponent, Props} from 'Popup/Components/ErrorBoundary';

function renderShallow({children, ...props}: Partial<Props & RouteComponentProps<{}>>): ShallowWrapper {
  const mockedRouteComponentProps = getMockRouteComponentProps({});
  const mockedClasses = {fullWidth: 'fullWidth', padding: 'padding'};

  return shallow(
    <ErrorBoundaryComponent
      {...mockedRouteComponentProps}
      classes={mockedClasses}
      {...props}
    >
      {children || <span>empty</span>}
    </ErrorBoundaryComponent>,
  );
}

describe('ErrorBoundary', () => {
  it('renders children', () => {
    expect(renderShallow({})).toMatchSnapshot();
  });

  it('renders error message when child throws error', () => {
    const rendered = renderShallow({});
    (rendered.instance() as ErrorBoundaryComponent)
      .componentDidCatch(new Error('testError'), {componentStack: 'testStack'});

    expect(rendered.update()).toMatchSnapshot();
  });

  describe('menu interaction', () => {
    let rendered: ShallowWrapper;
    let routeComponentProps: RouteComponentProps<{}>;
    beforeEach(() => {
      routeComponentProps = getMockRouteComponentProps({});
      rendered = renderShallow({
        ...routeComponentProps,
        children: <div id="test">test</div>,
      });
      rendered.setState({
        error: new Error('testError'),
        info: {componentStack: 'testStack'},
      });
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });

    it('does not show the error message initially', () => {
      expect(rendered.debug()).not.toContain('testError');
    });

    it('clicking showError shows the error message and stack', () => {
      rendered.find('.showError').simulate('click');
      expect(rendered.debug()).toContain('testError');
      expect(rendered.debug()).toContain('testStack');
    });

    it('clicking goBack clears the error and navigates back', () => {
      rendered.find('.goBack').simulate('click');
      expect(routeComponentProps.history.goBack).toHaveBeenCalled();
    });

    it('clicking openGithub opens github', () => {
      rendered.find('.openGithub').simulate('click');
      expect(browser.tabs.create).toHaveBeenCalledWith({url: 'https://github.com/PassB/passB/issues'});
    });
  });

});
