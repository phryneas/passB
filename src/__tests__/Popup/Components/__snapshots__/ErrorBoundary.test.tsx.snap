// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary menu interaction renders correctly 1`] = `
<div>
  <withStyles(Typography)
    gutterBottom={true}
    key="heading"
    type="title"
  >
    An error occured
  </withStyles(Typography)>
  <withStyles(Typography)>
    An unknown error has been encountered. The error message has been hidden as it may contain sensitive information. For troubleshooting choose one of the following:
  </withStyles(Typography)>
  <withStyles(List)>
    <withStyles(ListItem)
      button={true}
      className="showError"
      onClick={[Function]}
    >
      <withStyles(ListItemText)
        primary="reveal the error message"
      />
    </withStyles(ListItem)>
    <withStyles(ListItem)
      button={true}
      className="goBack"
      onClick={[Function]}
    >
      <withStyles(ListItemText)
        primary="go back and do something different"
      />
    </withStyles(ListItem)>
    <withStyles(ListItem)
      button={true}
      className="openGithub"
      onClick={[Function]}
    >
      <withStyles(ListItemText)
        primary="open an issue on github"
      />
    </withStyles(ListItem)>
  </withStyles(List)>
</div>
`;

exports[`ErrorBoundary renders children 1`] = `
<span>
  empty
</span>
`;

exports[`ErrorBoundary renders error message when child throws error 1`] = `
<div>
  <withStyles(Typography)
    gutterBottom={true}
    key="heading"
    type="title"
  >
    An error occured
  </withStyles(Typography)>
  <withStyles(Typography)>
    An unknown error has been encountered. The error message has been hidden as it may contain sensitive information. For troubleshooting choose one of the following:
  </withStyles(Typography)>
  <withStyles(List)>
    <withStyles(ListItem)
      button={true}
      className="showError"
      onClick={[Function]}
    >
      <withStyles(ListItemText)
        primary="reveal the error message"
      />
    </withStyles(ListItem)>
    <withStyles(ListItem)
      button={true}
      className="goBack"
      onClick={[Function]}
    >
      <withStyles(ListItemText)
        primary="go back and do something different"
      />
    </withStyles(ListItem)>
    <withStyles(ListItem)
      button={true}
      className="openGithub"
      onClick={[Function]}
    >
      <withStyles(ListItemText)
        primary="open an issue on github"
      />
    </withStyles(ListItem)>
  </withStyles(List)>
</div>
`;
